### PR TITLE
Mark skills

### PR DESCRIPTION
**Pull request by Alan-Penkar**
Updated mark and created goalside mark.

### CODE REVIEW

File: soccer/gameplay/skills/goalside_mark.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+        self.add_transition(behavior.Behavior.State.start,
+                            behavior.Behavior.State.running, lambda: True,
+                            "immediately")
+        self.kick_eval = robocup.KickEvaluator(main.system_state())
+
+    def execute_running(self):
+        #pylint: disable=no-member
+        #Skill does nothing if mark point isn't given AND the ball or robot to mark can't be found
+        self.kick_eval = robocup.KickEvaluator(main.system_state())
```
Comment (JNeiger) [inline soccer/gameplay/skills/goalside_mark.py]: I don't think you need to create a new kick eval object every frame. You can just use the same one and clear the excluded robots whenever

File: soccer/gameplay/skills/goalside_mark.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+                          constants.Field.PenaltyShortDist + offset))
+        right_seg = robocup.Segment(
+            robocup.Point(constants.Field.PenaltyLongDist / 2 + offset, 0),
+            robocup.Point(constants.Field.PenaltyLongDist / 2 + offset,
+                          constants.Field.PenaltyShortDist + offset))
+        top_seg = robocup.Segment(
+            robocup.Point(-constants.Field.PenaltyLongDist / 2,
+                          constants.Field.PenaltyShortDist + offset),
+            robocup.Point(constants.Field.PenaltyLongDist / 2,
+                          constants.Field.PenaltyShortDist + offset))
```
Comment (JNeiger) [inline soccer/gameplay/skills/goalside_mark.py]: I see why this is here, but I wish there was a better way to do this (Intersection of a segment/line with the goal box with some offset). If you are comfortable making some other changes to the C++ interface to allow for this that would be awesome. If not, I can add it sometime.

Basically, the [field](https://github.com/RoboJackets/robocup-software/blob/master/soccer/gameplay/constants.py#L34) in python refers back to it's C++ [counterpart](https://github.com/RoboJackets/robocup-software/blob/f0cf06638ee427563ca7162bbbd65809c30f59f2/common/Field_Dimensions.hpp#L47). Here, there is something called our goal zone shape which is a composite shape. In short a [composite shape](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/CompositeShape.hpp) is a list of shapes which are or'ed together. In old times (like a year or two ago), the defensive area was two arc's with a rectangle in the middle so this made sense. When we redid it for rectangles, we left the composite shape wrapper for some reason. 

Anyways, now, the composite shape just contains a single [rectangle](https://github.com/RoboJackets/robocup-software/blob/f0cf06638ee427563ca7162bbbd65809c30f59f2/common/Field_Dimensions.hpp#L127). The [rectangle](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/Rect.hpp) class doesn't allow for segment intersection, but that's a good thing to add in my mind. If we overload the `bool intersects` function and take in a segment, very close to how it's done in the [segment `intersect` function](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/Segment.hpp#L51), I think it'll work out pretty well. In terms of actual implementation, you should be able to follow the same way they do the [`nearPoint` function](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/Rect.cpp#L105) where they create the 4 segments and just call the segment functions. It's up to you, but I think returning both intersections would be the best bet and let the user figure out which to use (and maybe return the number of intersect points). Finally, to access in the python side, you'll have to add the function call to [this location in robocup-py](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/Rect.cpp#L105), but you'll have to follow the setup in the [segment intersect function](https://github.com/RoboJackets/robocup-software/blob/master/soccer/gameplay/robocup-py.cpp#L683) (and [here](https://github.com/RoboJackets/robocup-software/blob/master/soccer/gameplay/robocup-py.cpp#L242)) to fully deal with the weird python <-> C++ conversion issues. A test or two should be added into [this file](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/RectTest.cpp) also.

In the future, composite shape should probably be removed, and just a rectangle should be used instead, but that may be out of scope for this.

File: soccer/gameplay/skills/goalside_mark.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+        #Sets the position to mark as the given mark position 
+        #or robot position if no mark position is given
+        pos = self.robot.pos
+        mark_pos = self.mark_point if self.mark_point is not None else self.mark_robot.pos
+
+        #Find best shot point from threat
+        self.kick_eval.add_excluded_robot(self.robot) #FIX: Should we really exclude all of our robots but the goalie?
+        shot_pt, shot_score = self.kick_eval.eval_pt_to_our_goal(mark_pos)
+        self.kick_eval.excluded_robots.clear()
```
Comment (JNeiger) [inline soccer/gameplay/skills/goalside_mark.py]: I think these are just copy paste from the caller function?

File: soccer/gameplay/skills/goalside_mark.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+        
+        #Finds the line from the mark position to the shot point and creates a line between them
+        #removing the overlap with the ball on one side and robot on the other
+        #This assumes even with mark position parameter that there is a robot there to avoid
+        mark_line_dir = (mark_pos - shot_pt).normalized()
+        mark_line, shot_pt = self.get_mark_line()
+        print(mark_line)
```
Comment (JNeiger) [inline soccer/gameplay/skills/goalside_mark.py]: In the final version, prints should probably be removed so we don't spam the console too much

File: soccer/gameplay/skills/goalside_mark.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        #Sets target point to place on line defined by ratio- 
+        #   - 0 being close to ball, 1 close to mark pt
+        self._target_point = None
+        if mark_line_dist > self.mark_line_thresh:
+            self._target_point = mark_line.nearest_point(pos)
+        else:
+            if self.mark_point is None:
+                adjusted_mark_pos = mark_pos - mark_line_dir * 2 * constants.Robot.Radius
+            else:
+                adjusted_mark_pos = mark_pos - mark_line_dir * constants.Ball.Radius
```
Comment (JNeiger) [inline soccer/gameplay/skills/goalside_mark.py]: These two are also calculated already in the `get_mark_line()` function. It feels weird to re-calculate them

File: soccer/gameplay/skills/goalside_mark.py (hunk 2 above)
Comment (JNeiger) [reply soccer/gameplay/skills/goalside_mark.py]: Oh, and to add the offsets, you can use the [`expand`](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/Rect.hpp#L67) function of rectangles with the [mixx/maxx/miny/maxy](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/Rect.hpp#L73) to create the X meter offsets outside the given rectangle.

File: soccer/gameplay/skills/goalside_mark.py (hunk 1 above)
Comment (Alan-Penkar) [reply soccer/gameplay/skills/goalside_mark.py]: Good call- changed this trying to debug the kickevaluator and never changed it back

File: soccer/gameplay/skills/goalside_mark.py (hunk 2 above)
Comment (Alan-Penkar) [reply soccer/gameplay/skills/goalside_mark.py]: I'll give this a go - but expect to do a pretty heavy review on my C++ skills.

File: soccer/gameplay/skills/goalside_mark.py (hunk 3 above)
Comment (JNeiger) [reply soccer/gameplay/skills/goalside_mark.py]: It's not bad by itself, but if a big chunk of code is being copy pasted and the same calculations are being done, it's a good idea to find a way to merge them. That way if someone wants to make a tweak later on, they don't have to try and edit two places. I'll have to look at the code a little more to see if that's the case here, but it felt a little deja vu looking at the two functions

File: common/Geometry2d/Rect.cpp [hunk 6]
```diff
@@ -18,6 +18,111 @@ bool Rect::intersects(const Rect& other) const {
...
+        code |= BOTTOM;
+    else if (y > maxy())      // above the clip window
+        code |= TOP;
+
+    return code;
+}
+
+bool Rect::intersects_(const Segment& other) const{
+    //These constants are needed in here and in the CohenSutherlandCode method - is there
+    //a place where we should make these definitions globally?
```
Comment (JNeiger) [inline common/Geometry2d/Rect.cpp]: If you don't need them outside these two functions, putting them at the top of this file should be fine

File: common/Geometry2d/Rect.cpp [hunk 7]
```diff
@@ -18,6 +18,111 @@ bool Rect::intersects(const Rect& other) const {
...
+    const int INSIDE = 0; // 0000
+    const int LEFT = 1;   // 0001
+    const int RIGHT = 2;  // 0010
+    const int BOTTOM = 4; // 0100
+    const int TOP = 8;    // 1000
+
+    //Code aggressively borrowed from wikipedia entry Cohen-Sutherland Line Clipping
```
Comment (JNeiger) [inline common/Geometry2d/Rect.cpp]: I would just link it here

File: common/Geometry2d/Rect.cpp [hunk 8]
```diff
@@ -18,6 +18,111 @@ bool Rect::intersects(const Rect& other) const {
...
+        code |= RIGHT;
+    if (y < miny())           // below the clip window
+        code |= BOTTOM;
+    else if (y > maxy())      // above the clip window
+        code |= TOP;
+
+    return code;
+}
+
+bool Rect::intersects_(const Segment& other) const{
```
Comment (JNeiger) [inline common/Geometry2d/Rect.cpp]: You shouldn't need the underscore as C++ will allow method overloads depending on the arguments

File: common/Geometry2d/Rect.cpp [hunk 9]
```diff
@@ -18,6 +18,111 @@ bool Rect::intersects(const Rect& other) const {
     }
 }
 
+int Rect::CohenSutherlandCode(const Point& other) const{
```
Comment (JNeiger) [inline common/Geometry2d/Rect.cpp]: I'm on the fence about the naming for this function. At first glance it felt like just saying it's the Cohen Sutherland C++ code, but after looking at the algorithm, I see it's referring to the out code. Making it `CohenSutherlandOutCode`, `ComputeOutCode` etc may make it clearer.

File: common/Geometry2d/Rect.cpp [hunk 10]
```diff
@@ -18,6 +18,111 @@ bool Rect::intersects(const Rect& other) const {
...
+    const int INSIDE = 0; // 0000
+    const int LEFT = 1;   // 0001
+    const int RIGHT = 2;  // 0010
+    const int BOTTOM = 4; // 0100
+    const int TOP = 8;    // 1000
```
Comment (JNeiger) [inline common/Geometry2d/Rect.cpp]: You can use `0x01` to make this a hex literal or `1<<2` if you want to specify it's a bitwise field. Either of those makes it much clearer that you are using them in their binary form

File: common/Geometry2d/Rect.cpp [hunk 11]
```diff
@@ -18,6 +18,111 @@ bool Rect::intersects(const Rect& other) const {
...
+    double y0 = p0.y();
+    Point p1 = other.pt[1];
+    double x1 = p1.x();
+    double y1 = p1.y();
+
+    int outcode0 = CohenSutherlandCode(p0);
+    int outcode1 = CohenSutherlandCode(p1);
+    bool accept = false;
+
+    while (true) {
```
Comment (JNeiger) [inline common/Geometry2d/Rect.cpp]: There is no way this can become an infinite loop right? Even with a 0 by 0 rectangle and other possible edge cases?

File: common/Geometry2d/Rect.hpp [hunk 12]
```diff
@@ -76,6 +76,9 @@ class Rect : public Shape {
     bool nearSegment(const Segment& seg, float threshold) const;
 
     bool intersects(const Rect& other) const;
+    bool intersects_(const Segment& other) const;
+
+    int CohenSutherlandCode(const Point& other) const;
```
Comment (JNeiger) [inline common/Geometry2d/Rect.hpp]: This function can be private since it shouldn't be used anywhere else

File: common/Geometry2d/Rect.cpp [hunk 13]
```diff
@@ -18,6 +18,111 @@ bool Rect::intersects(const Rect& other) const {
...
+    const int BOTTOM = 4; // 0100
+    const int TOP = 8;    // 1000
+
+    //Code aggressively borrowed from wikipedia entry Cohen-Sutherland Line Clipping
+    Point p0 = other.pt[0];
+    double x0 = p0.x();
+    double y0 = p0.y();
+    Point p1 = other.pt[1];
+    double x1 = p1.x();
+    double y1 = p1.y();
```
Comment (JNeiger) [inline common/Geometry2d/Rect.cpp]: If you don't want to deal with writing `other.pt[1]` every single time, you can also do the `xmin/xmax/ymin/ymax` stuff too.

File: soccer/gameplay/skills/goalside_mark.py [hunk 14]
```diff
@@ -69,18 +62,14 @@ def execute_running(self):
         #   - 0 being close to ball, 1 close to mark pt
         self._target_point = None
         if mark_line_dist > self.mark_line_thresh:
-            self._target_point = mark_line.nearest_point(pos)
+            self._target_point = mark_line.nearest_point(self.robot.pos)
```
Comment (JNeiger) [inline soccer/gameplay/skills/goalside_mark.py]: self.robot can be none at times if we don't have enough robots to fill roles

File: soccer/gameplay/skills/goalside_mark.py [hunk 15]
```diff
@@ -114,6 +103,11 @@ def mark_point(self) -> robocup.Point:
     @mark_point.setter
     def mark_point(self, value: robocup.Point):
         self._mark_point = value
+    
+    # Sets the position to mark as the given mark position 
+    # or robot position if no mark position is given
+    def _reset_mark_pos(self):
+        self.mark_pos = self.mark_point if self.mark_point is not None else self._mark_robot.pos
```
Comment (JNeiger) [inline soccer/gameplay/skills/goalside_mark.py]: Is it ever possible that `mark_point` will be None and `_mark_robot` will be None?

**Comment by JNeiger** [comment]
Looking good so far!

For the tests, you can edit[ this file](https://github.com/RoboJackets/robocup-software/blob/master/common/Geometry2d/RectTest.cpp). The functions are in the form `TEST([Class], [test_name])`. There's been two ways I've been doing tests. The first is one test case for each specific function I want to test and the other is a test case for each edge case.

For Geometry2d I think we've been doing a test case for each function. Inside that we just test the function against a few "good" results to make sure it performs correct (Both successes and failures), then against every single edge case that it could hit.

In terms of the `EXPECT` stuff, there's TRUE, FALSE, NEAR, LT, GT, EQ are the big ones I use. This is GTest so there should be quite a few examples elsewhere.

You should be able to then run `make tests` to actually run the test suite. If you scroll up, you should see your function under the Rectangle class (IIRC).

File: common/Geometry2d/Rect.cpp [hunk 16]
```diff
@@ -18,6 +25,125 @@ bool Rect::intersects(const Rect& other) const {
...
+    }
+    return accept;
+}
+
+std::vector<Point> Rect::corners(){
+    std::vector<Point> tmp;
+    tmp.push_back(Point(minx(),miny()));
+    tmp.push_back(Point(minx(),maxy()));
+    tmp.push_back(Point(maxx(),maxy()));
+    tmp.push_back(Point(maxx(),miny()));
```
Comment (JNeiger) [inline common/Geometry2d/Rect.cpp]: If you do `tmp.emplace_back(minx(), miny())` it's more efficient since it doesn't have to create the Point object, then copy it into memory. It just creates it in memory

File: common/Geometry2d/Rect.hpp [hunk 17]
```diff
@@ -63,19 +63,61 @@ class Rect : public Shape {
...
     bool nearPoint(Point pt, float threshold) const override;
     bool nearSegment(const Segment& seg, float threshold) const;
 
     bool intersects(const Rect& other) const;
+    
+    /*
+    * Calculates intersection point(s) between the rectangle and a line segment
+    * Uses Cohen Sutherland line clipping algorithm
+    */
+    bool intersects_(const Segment& other, Point* intr1, Point* intr2) const;
```
Comment (JNeiger) [inline common/Geometry2d/Rect.hpp]: The `intr1` and `intr2` should default to `nullptr`

File: common/Geometry2d/Rect.cpp [hunk 18]
```diff
@@ -18,6 +25,125 @@ bool Rect::intersects(const Rect& other) const {
...
+    double y0 = p0.y();
+    Point p1 = other.pt[1];
+    double x1 = p1.x();
+    double y1 = p1.y();
+
+    int outcode0 = CohenSutherlandCode(p0);
+    int outcode1 = CohenSutherlandCode(p1);
+    
+    //intr1, intr2 = nullptr;
+    Point **nextPoint = &intr1;
```
Comment (JNeiger) [inline common/Geometry2d/Rect.cpp]: `intr1` can be a nullptr if it's not passed into the function. This would break the program. You should have two function variables if you really want to keep their state then just set intr1 and intr2 to their correct values if they are non-null at the end

Comment (JNeiger) [reply common/Geometry2d/Rect.cpp]: This wouldn't need to be a double pointer then. You could just make it a list and append it

File: common/Geometry2d/Rect.cpp [hunk 19]
```diff
@@ -18,6 +25,125 @@ bool Rect::intersects(const Rect& other) const {
...
+    tmp.push_back(Point(maxx(),miny()));
+    return tmp;
+}
+
+std::vector<Point> Rect::pointList(){
+    std::vector<Point> tmp;
+    tmp.push_back(pt[0]);
+    tmp.push_back(pt[1]);
+    return tmp;
+}
```
Comment (JNeiger) [inline common/Geometry2d/Rect.cpp]: I'm not sure if this is a good idea to add. It's already there through the `.pt[0]` and `.pt[1]` values. If it's for debugging and you were trying to print it out or something, it's better to just overload the `<<` operator and do it the correct way

Comment (JNeiger) [reply common/Geometry2d/Rect.cpp]: That can be a different pr. You don't have to add that

File: common/Geometry2d/Rect.cpp [hunk 20]
```diff
@@ -35,17 +161,33 @@ bool Rect::hit(const Segment& seg) const {
...
-    pt[1].y() = max(pt[1].y(), p.y());
+    float _minx = minx();
+    float _miny = miny();
+    float _maxx = maxx();
+    float _maxy = maxy();
+
+    pt[0].x() = min(_minx, (float)p.x());
+    pt[0].y() = min(_miny, (float)p.y());
+    pt[1].x() = max(_maxx, (float)p.x());
+    pt[1].y() = max(_maxy, (float)p.y());
```
Comment (JNeiger) [inline common/Geometry2d/Rect.cpp]: For this and the pad function, you should make it clear in the hpp that this changes which corners pt1 and pt2 are. This makes it bottom left and upper right always.

File: common/Field_Dimensions.hpp [hunk 21]
```diff
@@ -44,17 +44,41 @@ struct Field_Dimensions {
...
+    /*
+    * Provides a rect that is a padded version of their goalbox
+    * used mostly for movement at the play level
+    * exposed to python via constants.Field
+    */
+    Geometry2d::Rect TheirGoalZoneShapePadded(float padding){
+      Geometry2d::Rect tmp = Geometry2d::Rect(_TheirGoalZoneShape);
+      tmp.pad(padding);
+      return tmp;
+    };
```
Comment (JNeiger) [inline common/Field_Dimensions.hpp]: I'm on the fence about these two. This struct is built like it's just supposed to be the raw field information from the packet and make it very simple to pull that data. Additionally, this only saves one or two lines in python where it's used. 

The feature is nice, but is that enough reason to add it to a place it might not fit in

File: common/Geometry2d/Rect.hpp (hunk 17 above)
Comment (JNeiger) [inline common/Geometry2d/Rect.hpp]: This can just be named `intersects` and function overloading will determine which one they reach

File: soccer/gameplay/robocup-py.cpp [hunk 22]
```diff
@@ -251,6 +251,29 @@ boost::python::object Segment_segment_intersection(Geometry2d::Segment* self,
     }
 }
 
+boost::python::object Our_goalzone_padded(float padding){
+    Geometry2d::Rect tmp = Geometry2d::Rect(Field_Dimensions::Current_Dimensions.OurGoalZoneShape());
+    tmp.expand(Geometry2d::Point(-Field_Dimensions::Current_Dimensions.PenaltyLongDist()/2 - padding, Field_Dimensions::Current_Dimensions.PenaltyShortDist() + padding));
+    tmp.expand(Geometry2d::Point(Field_Dimensions::Current_Dimensions.PenaltyLongDist()/2 + padding, Field_Dimensions::Current_Dimensions.PenaltyShortDist() + padding));
+    return boost::python::object(tmp);
+}
```
Comment (JNeiger) [inline soccer/gameplay/robocup-py.cpp]: I think this is obe since you made a `pad` function

File: soccer/gameplay/robocup-py.cpp [hunk 23]
```diff
@@ -715,8 +741,12 @@ BOOST_PYTHON_MODULE(robocup) {
         .def("min_y", &Geometry2d::Rect::miny)
         .def("max_x", &Geometry2d::Rect::maxx)
         .def("max_y", &Geometry2d::Rect::maxy)
+        .def("corners", &Geometry2d::Rect::corners)
+        .def("points", &Geometry2d::Rect::pointList)
+        .def("pad", &Geometry2d::Rect::pad)
         .def("near_point", &Geometry2d::Rect::nearPoint)
         .def("intersects_rect", &Geometry2d::Rect::intersects)
+        .def("intersects_segment", &Rect_segment_intersection)
```
Comment (JNeiger) [inline soccer/gameplay/robocup-py.cpp]: Just to make the naming convention the same between pts, lines, and segments, this should flip. IE: `segment_intersection` instead of `intersects_segment`. Feel free to change up  `intersects_rect` as well since you need to change the function it points to. It'll need a wrapper similar to how you do `segment` intersection

File: soccer/gameplay/robocup-py.cpp [hunk 24]
```diff
@@ -681,6 +704,9 @@ BOOST_PYTHON_MODULE(robocup) {
         .def("nearly_equals", &Geometry2d::Point::nearlyEquals)
         .staticmethod("direction");
 
+    class_<std::vector<Geometry2d::Point>>("vector_Point")
+        .def(vector_indexing_suite<std::vector<Geometry2d::Point>>());
```
Comment (JNeiger) [inline soccer/gameplay/robocup-py.cpp]: If this is for the `corners` function, it might be reasonable to just create a wrapper and return and `boost::python::list`. It'll be less "weird" objects to deal with in python.

File: common/Geometry2d/Rect.hpp [hunk 25]
```diff
@@ -63,19 +63,61 @@ class Rect : public Shape {
...
     float maxy() const { return std::max(pt[0].y(), pt[1].y()); }
 
+    /*
+    * The corners() function lists the 4 corners of the rectangle
+    * in a predictable order regardless of the 2 corners defined on
+    * construction.
+    * BottomLeft, TopLeft, TopRight, BottomRight
+    * exposed to python as corners()
+    */
+    std::vector<Point> corners();
```
Comment (JNeiger) [inline common/Geometry2d/Rect.hpp]: Is this even being used anywhere?

File: soccer/gameplay/skills/goalside_mark.py [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+
+        self.add_transition(behavior.Behavior.State.start,
+                            behavior.Behavior.State.running, lambda: True,
+                            "immediately")
+        self.kick_eval = robocup.KickEvaluator(main.system_state())
+        self.mark_pos = None
+        self.adjusted_mark_pos = None
+
+    def execute_running(self):
+        #pylint: disable=no-member
```
Comment (JNeiger) [inline soccer/gameplay/skills/goalside_mark.py]: Which variable is giving the pylint problem?

File: soccer/gameplay/skills/goalside_mark.py [hunk 27]
```diff
@@ -0,0 +1,163 @@
...
+            self._target_point = self.adjusted_mark_pos - (
+                self.mark_pos -
+                shot_pt).normalized() * self.ratio * mark_line.length()
+
+        #Drawing for simulator
+        main.system_state().draw_circle(self.mark_pos, constants.Robot.Radius * 1.2,
+                                        (0, 127, 255), "Mark")
+
+        #Move robot into position and face the ball
+        self.robot.move_to(self._target_point)
```
Comment (JNeiger) [inline soccer/gameplay/skills/goalside_mark.py]: This should be a move subbehavior since its there. `SingleRobotCompositeBehavior` allows for a single subbehavior

File: common/Geometry2d/Rect.cpp (hunk 8 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.cpp]: it was getting f***ed up in boost and I didn't really want to figure out how to fix it....

File: common/Geometry2d/Rect.cpp (hunk 9 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.cpp]: That seems doable

Comment (Alan-Penkar) [reply common/Geometry2d/Rect.cpp]: refactored to CohenSutherlandOutCode

File: common/Geometry2d/Rect.cpp (hunk 10 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.cpp]: using hex literals

File: common/Geometry2d/Rect.cpp (hunk 11 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.cpp]: Fixed... GTest cases included in latest commit

File: common/Geometry2d/Rect.hpp (hunk 12 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.hpp]: changed

File: common/Geometry2d/Rect.cpp (hunk 13 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.cpp]: meh.

File: soccer/gameplay/skills/goalside_mark.py (hunk 14 above)
Comment (Alan-Penkar) [reply soccer/gameplay/skills/goalside_mark.py]: added self.robot is not None to the conditions required to execute_running

File: soccer/gameplay/skills/goalside_mark.py (hunk 15 above)
Comment (Alan-Penkar) [reply soccer/gameplay/skills/goalside_mark.py]: I mean... mark_robot could be a string.  But specifically stated that one of these is required in the docstring

File: common/Geometry2d/Rect.hpp (hunk 17 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.hpp]: no longer relevant

File: common/Geometry2d/Rect.cpp (hunk 16 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.cpp]: huh- how did that work exactly?  it's a Point vector, emplace is allowing me to add non-points?
(done)

File: common/Geometry2d/Rect.cpp (hunk 18 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.cpp]: done

File: common/Geometry2d/Rect.cpp (hunk 19 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.cpp]: removed

File: common/Geometry2d/Rect.cpp (hunk 20 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.cpp]: done

File: common/Geometry2d/Rect.hpp (hunk 17 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.hpp]: frankly... I just didn't want to deal with it in boost- how big a deal is this?

File: soccer/gameplay/robocup-py.cpp (hunk 22 above)
Comment (Alan-Penkar) [reply soccer/gameplay/robocup-py.cpp]: you're right - removed

File: soccer/gameplay/robocup-py.cpp (hunk 23 above)
Comment (Alan-Penkar) [reply soccer/gameplay/robocup-py.cpp]: I'm not sure why it needs a wrapper?  I just changed the name and it can call the same C++ function.

File: common/Geometry2d/Rect.hpp (hunk 25 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.hpp]: yes, recreationally

File: common/Geometry2d/Rect.cpp (hunk 16 above)
Comment (JNeiger) [reply common/Geometry2d/Rect.cpp]: You basically pass in the direct variables that usually go to the constructor directly into the `emplace_back`. It then forwards those to the real constructor

https://stackoverflow.com/questions/23717151/why-emplace-back-is-faster-than-push-back/23717223#23717223

File: common/Geometry2d/Rect.hpp (hunk 17 above)
Comment (JNeiger) [reply common/Geometry2d/Rect.hpp]: I think it's a pretty big deal, but not earth shattering. It would mess up the function overloading in C++ land. I made a patch for it since I was super familiar with how to do it. I don't think boost allows overloading for python functions so you have to build a second function, that calls the overloaded function.

Here's a link to the [patch](https://gist.github.com/JNeiger/96f2fa64c92cd0736d757dd4bd240122). Basically download the zip. Unzip it into the `robocup-software` directory. Run `git apply fix_underscore.patch` to apply the patch. You'll have to commit this afterwards then.

Feel free to ignore this and try it yourself if you would like, but it sounds like you've fully experienced all the boost::python you wanted to :P

File: soccer/gameplay/robocup-py.cpp (hunk 23 above)
Comment (JNeiger) [reply soccer/gameplay/robocup-py.cpp]: Long story short, the callbacks boost does to convert a python function call to a C++ function call doesn't really deal with types super well. I believe it uses the type definition of the target function to figure out what to convert the python objects to. If there are two definitions with the same name, it doesn't know which one to pick and dies. If you use the wrapper functions, it picks the correct one

File: soccer/gameplay/robocup-py.cpp (hunk 24 above)
Comment (Alan-Penkar) [reply soccer/gameplay/robocup-py.cpp]: done

File: soccer/gameplay/skills/goalside_mark.py (hunk 27 above)
Comment (Alan-Penkar) [reply soccer/gameplay/skills/goalside_mark.py]: This is not a composite behavior - I should change that?

File: common/Field_Dimensions.hpp (hunk 21 above)
Comment (Alan-Penkar) [reply common/Field_Dimensions.hpp]: so- I agree this doesn't totally fit in.  The reason I wanted it wrapped up here was because it directly changes the object you call it on in C++, so trying to do this in python leaves us open to someone accidentally changing the goalbox that gets called in all of python...  Tried it in the bridge file, but because we keep info in an instance of the class, that didn't work right.

File: common/Geometry2d/Rect.hpp (hunk 17 above)
Comment (Alan-Penkar) [reply common/Geometry2d/Rect.hpp]: ha- if you can't tell from the degradation of my commit messages, I think I'm done with C++ now and want to get back to the data driven stuff we've discussed.... trying so hard to close this out.  But I'll refactor this.

Comment (Alan-Penkar) [reply common/Geometry2d/Rect.hpp]: done

File: soccer/gameplay/robocup-py.cpp (hunk 23 above)
Comment (Alan-Penkar) [reply soccer/gameplay/robocup-py.cpp]: got it- it's required because of the overloading only.  It wouldn't have been with the underscore

File: soccer/gameplay/skills/goalside_mark.py (hunk 26 above)
Comment (Alan-Penkar) [reply soccer/gameplay/skills/goalside_mark.py]: this is the one thing I'm not sure how to deal with- don't know how pylint works or what its looking for

File: common/Field_Dimensions.hpp (hunk 21 above)
Comment (JNeiger) [reply common/Field_Dimensions.hpp]: I didn't even think about that. This is probably ok then

File: soccer/gameplay/skills/goalside_mark.py (hunk 27 above)
Comment (JNeiger) [reply soccer/gameplay/skills/goalside_mark.py]: Eh, I was looking at some of the other skills some more and it's like 50-50 whether they do this. It's up to you

File: common/Geometry2d/RectTest.cpp [hunk 28]
```diff
@@ -18,16 +20,63 @@ TEST(Rect, hit) {
...
+    EXPECT_EQ(intersectExample.CohenSutherlandOutCode(Point(1.5,2.5)), 0x0A);
+}*/
+
+TEST(Rect, degenerageCohenCodes){
+    EXPECT_EQ(nullExample.CohenSutherlandOutCode(Point(0,0)), 0x00);
+    EXPECT_EQ(nullExample.CohenSutherlandOutCode(Point(1,0)), 0x02);
+    EXPECT_EQ(nullExample.CohenSutherlandOutCode(Point(-1,0)), 0x01);
+    EXPECT_EQ(nullExample.CohenSutherlandOutCode(Point(0,1)), 0x08);
+    EXPECT_EQ(nullExample.CohenSutherlandOutCode(Point(0,-.1)), 0x04);
+}
```
Comment (JNeiger) [inline common/Geometry2d/RectTest.cpp]: This function needs to be commented out too since `CohenSutherlandOutCode` is private

File: soccer/gameplay/skills/goalside_mark.py (hunk 26 above)
Comment (JNeiger) [reply soccer/gameplay/skills/goalside_mark.py]: So I removed this and ran `make pylint`. There's a bunch of crap that comes up, but nothing similar to the `self.mark_robot` has no member `pos` that came up earlier. Try taking it out and seeing if circle ci blows up.

File: soccer/gameplay/robocup-py.cpp (hunk 24 above)
Comment (JNeiger) [reply soccer/gameplay/robocup-py.cpp]: This part can be removed now I think

File: common/Geometry2d/RectTest.cpp (hunk 28 above)
Comment (Alan-Penkar) [reply common/Geometry2d/RectTest.cpp]: Done

File: soccer/gameplay/skills/goalside_mark.py (hunk 26 above)
Comment (Alan-Penkar) [reply soccer/gameplay/skills/goalside_mark.py]: when I run make pylint - the only thing that comes up is related to a network util file that I've definitely never touched
